feature-libs/tracking/tms/core/services/tms.service.ts

```diff
@@ -0,0 +1,100 @@
+import { isPlatformServer } from '@angular/common';
+import {
+  Inject,
+  Injectable,
+  Injector,
+  isDevMode,
+  OnDestroy,
+  PLATFORM_ID,
+} from '@angular/core';
+import { CxEvent, EventService, WindowRef } from '@spartacus/core';
+import { merge, Observable, Subscription } from 'rxjs';
+import { TmsConfig } from '../config/tms-config';
+import { TmsCollector } from '../model/tms.model';
+
+/**
+ * This service interacts with the configured data layer object by pushing the Spartacus events to it.
+ */
+@Injectable({ providedIn: 'root' })
+export class TmsService implements OnDestroy {
+  /**
+   * Stores subscriptions to events.
+   */
+  protected subscription = new Subscription();
+
+  constructor(
+    protected eventsService: EventService,
+    protected windowRef: WindowRef,
+    protected tmsConfig: TmsConfig,
+    protected injector: Injector,
+    @Inject(PLATFORM_ID) protected platformId: any
+  ) {}
+
+  /**
+   * This method is called only once to start collecting and dispatching events to GTM
+   */
+  collect(): void {
+    if (isPlatformServer(this.platformId)) {
+      return;
+    }
+
+    for (const tmsCollectorConfig in this.tmsConfig.tagManager) {
+      const collectorConfig =
+        this.tmsConfig.tagManager[tmsCollectorConfig] ?? {};
+      const events =
+        collectorConfig.events?.map((event) => this.eventsService.get(event)) ||
+        [];
+
+      if (!collectorConfig.collector) {
```

Comment:
I would move this line before mapping the events (to not call eventService.get if not needed, as it's costly)